libp2p/src/builder.rs

```diff
@@ -0,0 +1,1795 @@
+                                (peer_id, futures::future::Either::Left(upgrade))
+                            }
+                            futures::future::Either::Right((peer_id, upgrade)) => {
+                                (peer_id, futures::future::Either::Right(upgrade))
+                            }
+                        },
+                    )
+                )
+            }
+            pub async fn without_noise(self) -> Result<SwarmBuilder<$providerCamelCase, BandwidthLoggingPhase<impl AuthenticatedMultiplexedTransport, R>>, WebsocketError> {
+                construct_behaviour_builder!(
+                    self,
+                    $dnsTcp,
+                    libp2p_tls::Config::new(&self.keypair).map_err(|e| WebsocketError(AuthenticationErrorInner::from(e).into()))?
+                )
+            }
+        }
+
+        #[cfg(all(not(target_arch = "wasm32"), feature = $providerKebabCase, feature = "dns", feature = "websocket", feature = "noise"))]
+        impl<T: AuthenticatedMultiplexedTransport, R>
+            SwarmBuilder<$providerCamelCase, WebsocketNoisePhase< T, R, WithoutTls>>
+        {
+            pub async fn with_noise(self) -> Result<SwarmBuilder<$providerCamelCase, BandwidthLoggingPhase<impl AuthenticatedMultiplexedTransport, R>>, WebsocketError> {
+                let _ = self.phase.tls_config;
+
+                construct_behaviour_builder!(
+                    self,
+                    $dnsTcp,
+                    libp2p_noise::Config::new(&self.keypair).map_err(|e| WebsocketError(AuthenticationErrorInner::from(e).into()))?
+                )
+            }
+        }
+    };
+}
+
+impl_websocket_noise_builder!(
+    "async-std",
+    AsyncStd,
+    libp2p_dns::DnsConfig::system(libp2p_tcp::async_io::Transport::new(
+        libp2p_tcp::Config::default(),
+    ))
+);
+// TODO: Unnecessary await for Tokio Websocket (i.e. tokio dns). Not ideal but don't know a better way.
+impl_websocket_noise_builder!(
+    "tokio",
+    Tokio,
+    futures::future::ready(libp2p_dns::TokioDnsConfig::system(
+        libp2p_tcp::tokio::Transport::new(libp2p_tcp::Config::default())
+    ))
+);
+
+#[derive(Debug, thiserror::Error)]
+#[error(transparent)]
+pub struct WebsocketError(WebsocketErrorInner);
+
+#[derive(Debug, thiserror::Error)]
+enum WebsocketErrorInner {
+    #[error("Dns")]
+    #[cfg(any(feature = "tls", feature = "noise"))]
+    Authentication(#[from] AuthenticationErrorInner),
+    #[cfg(feature = "dns")]
+    #[error("Dns")]
+    Dns(#[from] io::Error),
+}
+
+pub struct BandwidthLoggingPhase<T, R> {
+    relay_behaviour: R,
+    transport: T,
+}
+
+impl<T: AuthenticatedMultiplexedTransport, Provider, R>
+    SwarmBuilder<Provider, BandwidthLoggingPhase<T, R>>
+{
+    pub fn with_bandwidth_logging(
+        self,
+    ) -> (
+        SwarmBuilder<Provider, BehaviourPhase<impl AuthenticatedMultiplexedTransport, R>>,
+        Arc<BandwidthSinks>,
+    ) {
```

Comment:
I think one important property of builders is that you can chain them. Can we store the `BandwidthSinks` inside the phase and return a tuple in the final `build` method?

Reply:
I don't know a way how we can do this without introducing significant complexity into all subsequent phases.

`build` would need to be generic over the fact that the user called `with_bandwidth_logging` in previous steps.

I find it actually quite clean that `with_bandwidth_logging` returns both. I don't find it intuitive if I would get the `BandwidthSinks` as a final step. At that point I would have already forgotten what I am getting.

Reply:
At the moment, where we insert this matters, right?

Could we change the API such that the user has to pass in `Arc<BandwidthSinks>`? Then we could still chain the builder.

Reply:
We could do that. We would only need to make the two constructors public:

https://github.com/libp2p/rust-libp2p/blob/master/libp2p/src/bandwidth.rs#L49-L54

https://github.com/libp2p/rust-libp2p/blob/2c3d6ad6da8663f0bc5751138aa163988fadd8ab/libp2p/src/bandwidth.rs#L109-L116

Fine by me.

Reply:
Resolving here. @thomaseizinger please reopen in case you feel strongly about it. We can still do this in the future introducing a new method and deprecating the old.